### PR TITLE
eventsByTag and currentEventsByTag to use Offset

### PR DESCRIPTION
**Pull request by richardimaoka**
Refs #21283

As discussed in the above issue, I didn't implement other offset than long (Sequence) for LevelDB journal.

I did my best for javadsl.Offset and scaladsl.Offset, but not sure if that's the correct way... please let me know if I should do them differently.

### CODE REVIEW

**Comment by richard-imaoka** [comment]
Will look into test failure

**Comment by richard-imaoka** [comment]
As it is an experimental module, do we keep the binary compatibility? If we do, let me fix my code.
>[error] (akka-persistence-query-experimental/*:mimaReportBinaryIssues) akka-persistence-query-experimental: Binary compatibility check failed!

File: akka-persistence-query/src/main/scala/akka/persistence/query/scaladsl/CurrentEventsByTagQuery.scala [hunk 1]
```diff
@@ -17,7 +17,7 @@ trait CurrentEventsByTagQuery extends ReadJournal {
    * is completed immediately when it reaches the end of the "result set". Events that are
    * stored after the query is completed are not included in the event stream.
    */
-  def currentEventsByTag(tag: String, offset: Long): Source[EventEnvelope, NotUsed]
+  def currentEventsByTag(tag: String, offset: Offset): Source[EventEnvelope, NotUsed]
```
Comment (patriknw) [inline akka-persistence-query/src/main/scala/akka/persistence/query/scaladsl/CurrentEventsByTagQuery.scala]: One topic for discussion is how we shall introduce this change without breaking plugins.
We can change the API, since persistence-qurey is experimental, but I think it can cause unnecessary trouble to do that in a 2.4.x. When we release Akka 2.5 we can change the API before removing the experimental label.

One way would be to add `CurrentEventsByTagQuery2` with the new signatures. Then in 2.5 we make that the only trait (and rename it to `CurrentEventsByTagQuery`).

File: akka-persistence-query/src/main/scala/akka/persistence/query/javadsl/Offset.scala [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * Copyright (C) 2015-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.persistence.query.javadsl
+
+import java.util.UUID
+
+trait Offset {
+  def hashCode: Int
+  def asScala: akka.persistence.query.scaladsl.Offset
```
Comment (patriknw) [inline akka-persistence-query/src/main/scala/akka/persistence/query/javadsl/Offset.scala]: I don't think we need different classes for scaladsl and javadsl. We can place it in `akka.persistence.query` package.

File: akka-persistence-query/src/main/scala/akka/persistence/query/scaladsl/Offset.scala [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+  override def compare(that: Sequence): Int = value.compare(that.value)
+}
+
+final case class TimeBasedUUID(override val value: UUID) extends javadsl.TimeBasedUUID(value) with Offset with Ordered[TimeBasedUUID] {
+  if (value == null || value.version != 1) {
+    throw new IllegalArgumentException("UUID " + value + " is not a time-based UUID")
+  }
+
+  override def hashCode: Int = value.hashCode
```
Comment (patriknw) [inline akka-persistence-query/src/main/scala/akka/persistence/query/scaladsl/Offset.scala]: `equals` also

File: akka-persistence-query/src/main/scala/akka/persistence/query/scaladsl/CurrentEventsByTagQuery.scala (hunk 1 above)
Comment (richard-imaoka) [inline akka-persistence-query/src/main/scala/akka/persistence/query/scaladsl/CurrentEventsByTagQuery.scala]: Sure, added EventsByTagQuery2 and CurrentEventsByTagQuery2.

**Comment by richard-imaoka** [comment]
@patriknw thanks for the review - I updated this following your comments. Could you let me know if it looks ok now?

File: akka-persistence-query/src/main/scala/akka/persistence/query/Offset.scala [hunk 4]
```diff
@@ -0,0 +1,38 @@
+/**
+ * Copyright (C) 2015-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.persistence.query
+
+import java.util.UUID
+
+trait Offset {
+  def hashCode: Int
```
Comment (patriknw) [inline akka-persistence-query/src/main/scala/akka/persistence/query/Offset.scala]: why is this needed?

File: akka-persistence-query/src/main/scala/akka/persistence/query/Offset.scala [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+final case class Sequence(val value: Long) extends Offset with Ordered[Sequence] {
+
+  override def hashCode: Int = value.hashCode
+
+  override def toString: String = value.toString
```
Comment (patriknw) [inline akka-persistence-query/src/main/scala/akka/persistence/query/Offset.scala]: why override `hashCode` and `toString`? isn't the case class enough?

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-persistence-query/src/main/scala/akka/persistence/query/Offset.scala (hunk 4 above)
Comment (richard-imaoka) [inline akka-persistence-query/src/main/scala/akka/persistence/query/Offset.scala]: Ah I didn't realize case classes implement hasCode automatically. So I removed it in the new commit.

File: akka-persistence-query/src/main/scala/akka/persistence/query/Offset.scala (hunk 5 above)
Comment (richard-imaoka) [inline akka-persistence-query/src/main/scala/akka/persistence/query/Offset.scala]: Yes, that makes more sense. I didn't realize hasCode was provided by case class, and it looks case class's toString is better than what I did - removed them in the new commit.

**Comment by patriknw** [comment]
@johanandren please review, so that we have 2 lgtm

**Review by johanandren — APPROVED** [review]

**Comment by patriknw** [comment]
thanks @richard-imaoka
